server/pulp/server/controllers/content.py

```diff
@@ -128,113 +137,392 @@ def refresh_content_source(content_source_id=None):
+    downloaded and imported into permanent storage. Instances of this call should
+    *not* be cached between download tasks. Each download task will be issued
+    its own conduit instance that is scoped to that run alone.
 
-    :param catalog_entry: The catalog entry to queue a download task for.
-    :type  catalog_entry: pulp.server.db.model.LazyCatalogEntry
+    Instances of this class are thread-safe. Calls into this instance do not
+    have to be coordinated for thread safety, the instance will take care of it itself.
     """
-    tags = [
-        resource_tag(RESOURCE_CONTENT_UNIT_TYPE, catalog_entry.unit_id),
-        action_tag('download')
-    ]
-    _download_one.apply_async_with_reservation(
-        RESOURCE_CONTENT_UNIT_TYPE,
-        catalog_entry.unit_id,
-        [catalog_entry],
-        tags=tags)
+
+    def __init__(self, task_id):
+        """
+        :param task_id: identifies the task being performed
+        :type  task_id: str
+        """
+        StatusMixin.__init__(self, task_id, LazyStatusConduitException)
+
+    def __str__(self):
+        return 'LazyStatusConduit'
+
+
+class LazyUnitDownloadStep(Step, DownloadEventListener):
+    """
+    A Step that downloads all the given requests. The downloader is configured
+    to download from the Pulp Streamer components.
+
+    :ivar download_requests: The download requests the step will process.
+    :type download_requests: list of nectar.request.DownloadRequest
+    :ivar download_config:   The keyword args used to initialize the Nectar
+                             downloader configuration.
+    :type download_config:   dict
+    :ivar downloader:        The Nectar downloader used to fetch the requests.
+    :type downloader:        nectar.downloaders.threaded.HTTPThreadedDownloader
+    """
+
+    def __init__(self, step_type, step_description, lazy_status_conduit, download_requests):
+        """
+        Initializes a Step that downloads all the download requests provided.
+
+        :param lazy_status_conduit: Conduit used to update the task status.
+        :type  lazy_status_conduit: LazyStatusConduit
+        :param download_requests:   List of download requests to process.
+        :type  download_requests:   list of nectar.request.DownloadRequest
+        """
+        super(LazyUnitDownloadStep, self).__init__(
+            step_type=step_type,
+            status_conduit=lazy_status_conduit,
+        )
+        self.description = step_description
+        self.download_requests = download_requests
+        self.download_config = {
+            MAX_CONCURRENT: int(pulp_conf.get('lazy', 'download_concurrency')),
+            HEADERS: {PULP_STREAM_REQUEST_HEADER: 'true'},
+            SSL_VALIDATION: True
+        }
+        self.downloader = HTTPThreadedDownloader(
+            DownloaderConfig(**self.download_config),
+            self
+        )
+
+    def _process_block(self, item=None):
```

Comment:
The new in-line docs in the step module help clarify things, and probably didn't exist when you wrote this.

This method shouldn't be overridden, and instead you should override the ``process_main()`` method to do this work.